modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequestBodyStream.java

```diff
@@ -66,14 +67,19 @@ public void addTracingHandler(ChunkHandler chunkHandler) {
     public void next() {
         assert closing == false : "cannot request next chunk on closing stream";
         assert handler != null : "handler must be set before requesting next chunk";
-        channel.eventLoop().submit(() -> {
+        final Future<?> future = channel.eventLoop().submit(() -> {
             requested = true;
             if (buf == null) {
                 channel.read();
             } else {
                 send();
             }
         });
+        future.addListener(f -> {
+            if (f.isSuccess() == false) {
+                channel.eventLoop().submit(() -> channel.pipeline().fireExceptionCaught(f.cause()));
```

Comment:
nit, try-catch is less verbose

```
                try {
                    send();
                } catch (Throwable t) {
                    channel.pipeline().fireExceptionCaught(t);
                }
```

Reply:
I think the try-catch form only covers when there is already a `buf` to send? If the stream needs to call `channel.read` first, the code will go through a different path to `send`. So adding a listener seems more comprehensive to me though a bit verbose (a somewhat inherent issue with listener based code).

Reply:
This `send()` is the only place so far that does not fire exception in pipeline. `channel.read()` and another `send()`, both happen in the pipeline and will trigger exception.

Reply:
Also you would add a listener on every `next()` call, possibly every 8kb of data. Once downstream handler start processing it's unlikely that send() will be invoked, because `buf` will be empty.

Reply:
Thanks for explaining! Those are good points. Pushed f40c2dce0b655f8c3115ee52a57260ef71753d79 as you suggested.